File: src/ts/longestStrChain.ts
namespace Version0 {
  export function dfs (word: string, wordSet: Set<string>, dp: Map<string, number>): number {
    if (dp.has(word)) {
      return dp.get(word)!
    }

    let result = 1
    for (let i = 0; i < word.length; i += 1) {
      const predecessor = word.substring(0, i) + word.substring(i + 1)
      if (wordSet.has(predecessor)) {
        result = Math.max(result, 1 + dfs(predecessor, wordSet, dp))
      }
    }
    dp.set(word, result)
    return result
  }
}

export function longestStrChain (words: string[]): number {
  let result = 0
  const wordSet = new Set(words)
  const dp = new Map<string, number>()

  for (const word of words) {
    result = Math.max(result, Version0.dfs(word, wordSet, dp))
  }

  return result
}

namespace Version1 {
  export function dfs (word: string, dp: Map<string, number>): number {
    const memo = dp.get(word) as number
    if (memo > 0) {
      return memo
    }

    let result = 1
    for (let i = 0; i < word.length; i += 1) {
      const predecessor = word.substring(0, i) + word.substring(i + 1)
      if (dp.has(predecessor)) {
        result = Math.max(result, 1 + dfs(predecessor, dp))
      }
    }
    dp.set(word, result)
    return result
  }
}

export function longestStrChain1 (words: string[]): number {
  let result = 0
  const dp = new Map<string, number>()
  for (const word of words) {
    dp.set(word, -1)
  }

  for (const word of words) {
    result = Math.max(result, Version1.dfs(word, dp))
  }

  return result
}

export function longestStrChain2 (words: string[]): number {
  words.sort((a, b) => a.length - b.length)
  let result = 0
  const dp = new Map<string, number>()

  for (const word of words) {
    let curLen = 1
    for (let i = 0; i < word.length; i += 1) {
      const predecessor = word.substring(0, i) + word.substring(i + 1)
      const preLen = dp.get(predecessor) ?? Number.MIN_SAFE_INTEGER
      curLen = Math.max(preLen + 1, curLen)
    }
    dp.set(word, curLen)
    result = Math.max(result, curLen)
  }

  return result
}

File: src/rs/longest_str_chain.rs
use std::collections::HashMap;

pub struct LongestStrChain {}
impl LongestStrChain {
    pub fn longest_str_chain(words: Vec<String>) -> i32 {
        let mut result = 0;
        let mut dp = HashMap::<String, i32>::new();
        for word in words.iter() {
            dp.insert(word.clone(), -1);
        }

        for word in words.iter() {
            result = result.max(Self::dfs(word.clone(), &mut dp))
        }

        result
    }

    fn dfs(word: String, dp: &mut HashMap<String, i32>) -> i32 {
        if let Some(memo) = dp.get(&word) {
            if *memo > 0 {
                return *memo;
            }
        }

        let mut result = 1;
        for i in 0..word.len() {
            let predecessor = String::new() + &word[0..i] + &word[i + 1..];
            if let Some(_) = dp.get(&predecessor) {
                result = result.max(1 + Self::dfs(predecessor, dp));
            }
        }

        dp.insert(word, result);
        result
    }

    pub fn longest_str_chain1(words: Vec<String>) -> i32 {
        let mut result = 0;
        let mut words = words;
        words.sort_unstable_by_key(|k| k.len());

        let mut dp = HashMap::<&String, i32>::new();
        for word in words.iter() {
            let mut cur_len = 1;
            for i in 0..word.len() {
                let predecessor = String::new() + &word[0..i] + &word[i + 1..];
                cur_len = if let Some(pre_len) = dp.get(&predecessor) {
                    cur_len.max(pre_len + 1)
                } else {
                    cur_len
                };
            }

            dp.insert(word, cur_len);
            result = result.max(cur_len);
        }

        result
    }

    pub fn longest_str_chain2(words: Vec<String>) -> i32 {
        let mut result = 0;
        let mut words = words;
        words.sort_unstable_by_key(|k| k.len());

        let mut dp = HashMap::<&String, i32>::new();
        for word in words.iter() {
            let mut cur_len = 1;
            for i in 0..word.len() {
                let mut predecessor = word.clone();
                predecessor.remove(i);
                cur_len = if let Some(pre_len) = dp.get(&predecessor) {
                    cur_len.max(pre_len + 1)
                } else {
                    cur_len
                };
            }

            dp.insert(word, cur_len);
            result = result.max(cur_len);
        }

        result
    }

    pub fn longest_str_chain3(words: Vec<String>) -> i32 {
        let mut result = 0;
        let mut words = words;
        words.sort_unstable_by_key(|k| k.len());

        let mut dp = HashMap::<&String, i32>::new();
        for word in words.iter() {
            let mut cur_len = 1;
            for i in 0..word.len() {
                let mut predecessor = word.clone();
                predecessor.remove(i);
                if let Some(pre_len) = dp.get(&predecessor) {
                    cur_len = cur_len.max(pre_len + 1);
                }
            }

            dp.insert(word, cur_len);
            result = result.max(cur_len);
        }

        result
    }
}
